### PR TITLE
Basic Vite setup

### PR DESCRIPTION
**Pull request by joshuapease**
### Overview

- #1 

What this PR does.
- Bare bones Vite config 
- DDEV setup for ports & daemons  
- Working webfonts
- Working JS 
- Avoids flash of unstyled content while in dev mode. 

What this PR (intentionally) doesn't do
- Set up Tailwind
- Set up Prettier & other coding standards
- Static asset copying (think icon SVGs, will address in future PR)

### How to test

Create a directory for your new Craft project

```bash
mkdir my-craft-project

# Then move into your directory
cd my-craft-project
```

Run the following `composer create-project` command. 

It uses this branch as the starter project.

```
docker run --rm -it -v "$PWD":/app -v ${COMPOSER_HOME:-$HOME/.composer}:/tmp composer create-project viget/craft-site-starter=dev-jp/vite-setup ./ --ignore-platform-reqs
```

After completing the setup steps (it prompts for your project's name), run the following commands.

👇 These will gradually become more automated as this project progresses.

```bash
# Start DDEV
ddev start

# Install Craft
ddev craft install

# Install Vite plugin
ddev craft plugin/install vite

# Launch the site!
ddev launch
```

What you should see:
- A styled page 
- Webfonts being loaded in the network tab.
- A console log (and no errors)

ℹ️  You may need a `ddev restart` to get rid of 404s around webfonts. The Vite config is looking for the `PRIMARY_SITE_URL` environment variable. It might not be fully loaded by DDEV. 

   

CleanShot 2024-05-14 at 09 34 26@2x

### CODE REVIEW

File: config/vite.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// you to visit the http or https port. If you plant to only
+// use http, make sure that PRIMARY_SITE_URL is set to http
+//
+// Match ports to .ddev/config.yaml -> web_extra_exposed_ports
+$host = Craft::$app->getRequest()->getIsConsoleRequest()
+    ? App::env('PRIMARY_SITE_URL')
+    : Craft::$app->getRequest()->getHostInfo();
+$httpPort = 3000;
+$httpsPort = 3001;
+$devServerPort = str_starts_with($host, 'https') ? $httpsPort : $httpPort;
```
Comment (joshuapease) [inline config/vite.php:16]: For a while I was having difficulty getting `https` working with DDEV (I've got it working great now). 

Figured it wasn't too much extra code to accommodate other's who might hit that issue.

File: src/css/app.css [hunk 2]
```diff
@@ -0,0 +1,41 @@
+@font-face {
+    font-display: swap;
+    font-family: 'Source Sans Pro';
+    font-style: normal;
+    font-weight: 300;
+    src: url('../fonts/sourcesanspro-light-webfont.woff2') format('woff2');
```
Comment (joshuapease) [inline src/css/app.css:6]: Vite can automatically copy & hash filenames for fonts. It just has to be a relative path from the css file to the font file.

File: templates/_layouts/base.twig [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+
+    <title>{{ siteName }}</title>
+
+    {# Preload Webfonts #}
+    <link rel="preload" href="{{ craft.vite.asset("src/fonts/sourcesanspro-bold-webfont.woff2") }}" as="font" type="font/woff2" crossorigin>
+    <link rel="preload" href="{{ craft.vite.asset("src/fonts/sourcesanspro-light-webfont.woff2") }}" as="font" type="font/woff2" crossorigin>
+    <link rel="preload" href="{{ craft.vite.asset("src/fonts/sourcesanspro-regular-webfont.woff2") }}" as="font" type="font/woff2" crossorigin>
+    <link rel="preload" href="{{ craft.vite.asset("src/fonts/sourcesanspro-semibold-webfont.woff2") }}" as="font" type="font/woff2" crossorigin>
```
Comment (joshuapease) [inline templates/_layouts/base.twig:13]: Figured this would be a nice example/reminder to preload your fonts.

File: vite.config.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                input: {
+                    app: 'src/js/app.js',
+                }
+            },
+        },
+        server: {
+            host: '0.0.0.0',
+            strictPort: true,
+            port: HTTP_PORT,
+            origin: env.PRIMARY_SITE_URL + ':' + originPort,
```
Comment (joshuapease) [inline vite.config.js:26]: This line is also needed to get webfonts, images, etc to work properly in dev mode.

You have to tell Vite the precise URL to use for any files referenced in your CSS & JS files.

**Review by maxfenton — APPROVED** [review]
Great work!

**Review by JJimmyFlynn — APPROVED** [review]

**Review by ten1seven — APPROVED** [review]

**Pull request merged**